source/collections/HashDictionary.ooc

```diff
@@ -1,91 +1,100 @@
 import structs/[HashBag, HashMap, ArrayList]
 
 HashDictionary: class {
-	_myHashBag: HashBag
+	_hashBag: HashBag
 	_capacity: Int
+	count ::= this _hashBag size()
+	empty ::= this _hashBag empty?()
+	
 	init: func { init ~withCapacity(10) }
 	init: func ~withCapacity (=_capacity) {
-		_myHashBag = HashBag new(this _capacity)
+		this _hashBag = HashBag new(this _capacity)
 	}
 	init: func ~copy (other: This) {
-		hashMapClone := other _myHashBag myMap clone()
-		_myHashBag = HashBag new(other _capacity)
-		this _myHashBag myMap = hashMapClone
+		hashMapClone := other _hashBag myMap clone()
+		this _hashBag = HashBag new(other _capacity)
+		this _hashBag myMap = hashMapClone
+	}
+	free: override func {
+		free(this _hashBag myMap keys data)
+		free(this _hashBag myMap keys)
+		for (i in 0 .. (this _hashBag myMap capacity)) {
+			next := (this _hashBag myMap buckets[i])
+			next dispose()
+		}
+		free(this _hashBag myMap buckets data)
+		free(this _hashBag myMap)
+		free(this _hashBag)
+		super()
 	}
 	clone: func -> This {
 		result := This new()
-		hashMapClone := this _myHashBag myMap clone()
-		result _myHashBag myMap = hashMapClone
+		hashMapClone := this _hashBag myMap clone()
+		result _hashBag myMap = hashMapClone
 		result
 	}
 	merge: func (other: This) -> This {
 		result := this clone()
-	result _myHashBag myMap merge!(other _myHashBag myMap)
+		result _hashBag myMap merge!(other _hashBag myMap)
 		result
 	}
-	/* WARNING: The defaultValue parameter for Covers must be new Cell(cover) */
 	get: func <T> (key: String, defaultValue: T) -> T {
 		result := defaultValue
-		if (_myHashBag contains?(key)) {
-			storedType := _myHashBag getClass(key)
-			// is `T` a derived type or the same type as the stored type?
-			if (T inheritsFrom?(storedType)) {
-				result = _myHashBag getEntry(key, T) value as T
+		if (_hashBag contains?(key)) {
+			storedType := this _hashBag getClass(key)
+			entryValue := this _hashBag getEntry(key, storedType) value
+			if (storedType inheritsFrom?(Cell)) {
+				entryValueCell := (entryValue as Cell<T>*)@
+				if (T inheritsFrom?(entryValueCell type))
+					result = entryValueCell get()
 			}
+			else if (T inheritsFrom?(storedType))
+				result = this _hashBag getEntry(key, T) value as T
 		}
 		result
 	}
-	/* WARNING: The Class parameter for Covers must be Cell<Cover> */
-	getAsType: func <T>(key: String, T: Class) -> T {
-		result := null
-		if (_myHashBag contains?(key)) {
-			storedType := _myHashBag getClass(key)
-			// is `T` a derived type or the same type as the stored type?
-			if (T inheritsFrom?(storedType))
-				result = _myHashBag getEntry(key, T) value as T
+	getAsType: func <T>(key: String, T: Class, defaultValue: T) -> T {
```

Comment:
What should I fix? It doesn't store the type in `Cell` anymore. The reason we specify `T` explicitly here is so that rock can understand the expected type. As mentioned previously, this is what makes `Future start()` work.